lib/elixir/lib/record.ex

```diff
@@ -397,35 +389,85 @@ defmodule Record do
       raise ArgumentError, "cannot invoke update style macro inside match"
     end
 
-    keyword = apply_underscore(fields, keyword)
+    case build_updates(keyword, fields, [], [], []) do
+      {updates, [], []} ->
+        build_update(updates, var)
 
-    Enum.reduce(keyword, var, fn {key, value}, acc ->
-      index = find_index(fields, key, 0)
+      {updates, vars, exprs} ->
 
-      if index do
         quote do
-          :erlang.setelement(unquote(index), unquote(acc), unquote(value))
+          {unquote_splicing(:lists.reverse(vars))} = {unquote_splicing(:lists.reverse(exprs))}
+          unquote(build_update(updates, var))
         end
-      else
+
+      {:error, key} ->
         raise ArgumentError, "record #{inspect(tag)} does not have the key: #{inspect(key)}"
-      end
+    end
+  end
+
+  defp build_update(updates, initial) do
+    updates
+    |> Enum.sort(fn {left, _}, {right, _} -> right <= left end)
+    |> Enum.reduce(initial, fn {key, value}, acc ->
+      quote(do: :erlang.setelement(unquote(key), unquote(acc), unquote(value)))
     end)
   end
 
-  # Gets a record key from the given var.
-  defp get(tag, fields, var, key) do
-    index = find_index(fields, key, 0)
+  defp build_updates([{:_, value} | rest], fields, updates, vars, exprs) do
+    {updates, vars, exprs} = build_updates(rest, fields, updates, vars, exprs)
 
-    if index do
-      quote do
-        :erlang.element(unquote(index), unquote(var))
+    if simple_argument?(value) do
+      updates = add_defaults(fields, updates, value, 2)
```

Comment:
If you add this test to the suit:
```
record = user(user(), _: :_, name: "meg", age: 12)
assert user(record, :name) == "meg"
assert user(record, :age) == 12
```

You will have the following error:
```
** (FunctionClauseError) no function clause matching in Record.add_defaults/4    
    
    The following arguments were given to Record.add_defaults/4:
    
        # 1
        []
    
        # 2
        [{2, "meg"}]
    
        # 3
        :_
    
        # 4
        4
    
    Attempted function clauses (showing 3 out of 3):
    
        defp add_defaults([_ | fields], [{i, _} = update | updates], value, i)
        defp add_defaults([_ | fields], updates, value, i)
        defp add_defaults([], [], _, _)
    
    (elixir) lib/record.ex:444: Record.add_defaults/4
    (elixir) lib/record.ex:445: Record.add_defaults/4
    (elixir) lib/record.ex:449: Record.add_defaults/4
    (elixir) lib/record.ex:420: Record.build_updates/5
    (elixir) lib/record.ex:392: Record.update/5
    expanding macro: RecordTest.user/2
    lib/elixir/test/elixir/record_test.exs:134: RecordTest."test records with default values"/1
```

Also, when the `:_` comes at the end of the keyword list or when you have repeated keys, it will execute some useless code. I believe it doesn't belong to this PR, but it would be nice if repeated keys we had some kind of warning.